data_safe_haven/external/api/graph_api.py

```diff
@@ -573,12 +593,145 @@ def get_id_from_username(self, username: str) -> str | None:
+                    application["principalDisplayName"] == application_name
+                ):
+                    self.logger.debug(
+                        f"Application role '[green]{application_role_name}[/]' already assigned to '{application_name}'.",
+                    )
+                    return
+            # Otherwise grant permissions for this role to the application
+            self.logger.debug(
+                f"Assigning application role '[green]{application_role_name}[/]' to '{application_name}'...",
+            )
+            request_json = {
+                "principalId": application_sp["id"],
+                "resourceId": microsoft_graph_sp["id"],
+                "appRoleId": app_role_id,
+            }
+            response = self.http_post(
+                f"{self.base_endpoint}/servicePrincipals/{microsoft_graph_sp['id']}/appRoleAssignments",
+                json=request_json,
+            )
+            self.logger.info(
+                f"Assigned application role '[green]{application_role_name}[/]' to '{application_name}'.",
+            )
+        except Exception as exc:
+            msg = f"Could not assign application role '{application_role_name}' to application '{application_name}'.\n{exc}"
+            raise DataSafeHavenMicrosoftGraphError(msg) from exc
+
+    def grant_delegated_role_permissions(
+        self, application_name: str, application_role_name: str
+    ) -> None:
+        """
+        Grant permissions for a particular role to an application.
+        See https://learn.microsoft.com/en-us/graph/permissions-grant-via-msgraph
+
+        Raises:
+            DataSafeHavenMicrosoftGraphError if the secret could not be created or already exists
+        """
+        try:
+            # Get service principals for Microsoft Graph and this application
+            microsoft_graph_sp = self.get_service_principal_by_name("Microsoft Graph")
+            if not microsoft_graph_sp:
+                msg = "Could not find Microsoft Graph service principal."
+                raise DataSafeHavenMicrosoftGraphError(msg)
+            application_sp = self.get_service_principal_by_name(application_name)
+            if not application_sp:
+                msg = "Could not find application service principal."
+                raise DataSafeHavenMicrosoftGraphError(msg)
+            # Check existing permissions
+            response = self.http_get(f"{self.base_endpoint}/oauth2PermissionGrants")
+            self.logger.debug(
+                f"Assigning delegated role '[green]{application_role_name}[/]' to '{application_name}'...",
+            )
+            # If there are existing permissions then we need to patch
+            application = next(
+                (
+                    app
+                    for app in response.json().get("value", [])
+                    if app["clientId"] == application_sp["id"]
+                ),
+                None,
+            )
```

Comment:
Will this only ever get the first item in the generator? Is that correct?

I don't think I understand what is happening here.

Reply:
It's the equivalent of doing:

```
filtered = [x for x in <sequence> if <condition>]
application = filtered[0] if filtered else None
```

Reply:
And we only ever need the first item here?

Reply:
Yes. We're trying to get the application with `clientId == application_sp["id"]` and there should only be one (or zero).

Reply:
Great, in that case maybe add a comment to explain that.

I think it would be easy to assume it is a mistake (at least I thought it might be :smile:).

Reply:
Done